ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -389,38 +221,215 @@ nullNetworkServerTracers = NetworkServerTracers {
       nstErrorPolicyTracer = nullTracer
     }
 
-
--- | Mutable state maintained by the network component.
+-- | Domain-specific rejection type. Only incoming connections can be rejected.
+-- Outgoing connections can still fail, but there is no "normal"
+-- (non-exceptional) reason to reject one.
 --
-data NetworkMutableState = NetworkMutableState {
-    nmsConnectionTable :: ConnectionTable IO Socket.SockAddr,
-    -- ^ 'ConnectionTable' which maintains information about current upstream and
-    -- downstream connections.
-    nmsPeerStates      :: StrictTVar IO (PeerStates IO Socket.SockAddr)
-    -- ^ 'PeerStates' which maintains state of each downstream / upstream peer
-    -- that errored, misbehaved or was not interesting to us.
-  }
-
-newNetworkMutableStateSTM :: STM.STM NetworkMutableState
-newNetworkMutableStateSTM =
-    NetworkMutableState <$> newConnectionTableSTM
-                        <*> newPeerStatesVarSTM
-
-newNetworkMutableState :: IO NetworkMutableState
-newNetworkMutableState = atomically newNetworkMutableStateSTM
+-- TODO constructor should include an explanation
+data RejectConnection (p :: Provenance) where
+  Rejected :: RejectConnection Remote
 
--- | Clean 'PeerStates' within 'NetworkMutableState' every 200s
+data SomeVersionedApplication ptcl vNumber vDataT provenance where
+  SomeVersionedResponderApp
+    :: ( Mx.HasResponder appType ~ True )
+    => NetworkServerTracers ptcl vNumber
+    -> Versions vNumber vDataT (OuroborosApplication appType ConnectionId ptcl IO BL.ByteString a b)
+    -> SomeVersionedApplication ptcl vNumber vDataT Remote
+  SomeVersionedInitiatorApp
+    :: ( Mx.HasInitiator appType ~ True )
+    => NetworkConnectTracers ptcl vNumber
+    -> Versions vNumber vDataT (OuroborosApplication appType ConnectionId ptcl IO BL.ByteString a b)
+    -> SomeVersionedApplication ptcl vNumber vDataT Local
+
+-- | Contains everything needed to run a mux'd versioned ouroboros application
+-- as initiator or responder. See use in `withConnections`. The idea is that
+-- the user will give a custom `request :: Provenance -> Type` type and must
+-- create a `ConnectionData ptcl vNumber provenance` with the same provenance
+-- as the request. In this way, the choice of `request` GADT can determine
+-- whether initiator or responder are even allowed (node-to-client server, for
+-- instance, does not allow initiation).
+data ConnectionData ptcl vNumber provenance where
+  -- | Locally-initiated connection data.
+  ConnectionDataLocal
```

Comment:
Why `ConnectionDataLoca` does not require `ErrorPolicies`?

Reply:
AFAICT the locally-initiated connections didn't use error policies before...

Reply:
Or maybe it's just that it wasn't at all clear that they did? Maybe it was the subscription modules which attached the error policies?